odl/tomo/data/uncompr_bin.py

```diff
@@ -0,0 +1,716 @@
+
+        Parameters
+        ----------
+        dstart : int, optional
+            Offset in bytes of the data field. By default, it is taken to
+            be the header size as determined from reading the header.
+            Backwards indexing with negative values is also supported.
+            Use a value larger than the header size to extract a data subset.
+        dend : int, optional
+            End position in bytes until which data is read (exclusive).
+            Backwards indexing with negative values  is also supported.
+            Use a value different from the file size to extract a data subset.
+        reshape_order : {'C', 'F', 'A'}, optional
+            Value passed as ``order`` parameter to `numpy.reshape`.
+            Reshaping is only done in case the whole data block is read.
+
+        Returns
+        -------
+        data : `numpy.ndarray`
+            The data read from `file`.
+
+        See Also
+        --------
+        read_header
+        """
+        self.file.seek(0, 2)  # 2 means "from the end"
+        filesize_bytes = self.file.tell()
+        if dstart is None:
+            dstart_abs = int(self.header_size)
+        elif dstart < 0:
+            dstart_abs = filesize_bytes + int(dstart)
+        else:
+            dstart_abs = int(dstart)
+
+        if dend is None:
+            dend_abs = int(filesize_bytes)
+        elif dend < 0:
+            dend_abs = int(dend) + filesize_bytes
+        else:
+            dend_abs = int(dend)
+
+        if dstart_abs >= dend_abs:
+            raise ValueError('invalid `dstart` and `dend`, resulting in '
+                             'absolute `dstart` >= `dend` ({} >= {})'
+                             ''.format(dstart_abs, dend_abs))
+        if dstart_abs < self.header_size:
+            raise ValueError('invalid `dstart`, resulting in absolute '
+                             '`dstart` < `header_size` ({} < {})'
+                             ''.format(dstart_abs, self.header_size))
+        if dend_abs > filesize_bytes:
+            raise ValueError('invalid `dend`, resulting in absolute '
+                             '`dend` > `filesize_bytes` ({} < {})'
+                             ''.format(dend_abs, filesize_bytes))
+
+        num_elems = (dend_abs - dstart_abs) / self.data_dtype.itemsize
+        if num_elems != int(num_elems):
+            raise ValueError('trying to read {} bytes, corresponding to '
+                             '{} elements of type {}'
+                             ''.format(dend_abs - dstart_abs, num_elems,
+                                       self.data_dtype))
+        self.file.seek(dstart_abs)
+        # Numpy determines byte order by itself
+        data = np.fromfile(self.file, dtype=self.data_dtype,
+                           count=int(num_elems))
+
+        if dstart_abs == self.header_size and dend_abs == filesize_bytes:
+            # Full dataset read, reshape to stored shape.
```

Comment:
Wait, this is NOT obvious. Sure, the doc mentions it but i guarantee that this will catch 98% of users of guard anyway (mostly because the parameter is an optional parameter and they may not even be using it). No other way to solve this?